cache/network/download_s3.go

```diff
@@ -0,0 +1,193 @@
+	}
+
+	if params.Bucket == "" {
+		return "", fmt.Errorf("bucket must not be empty")
+	}
+
+	cfg, err := loadAWSCredentials(
+		ctx,
+		params.Region,
+		params.AccessKeyID,
+		params.SecretAccessKey,
+		logger,
+	)
+	if err != nil {
+		return "", fmt.Errorf("load aws credentials: %w", err)
+	}
+
+	client := s3.NewFromConfig(*cfg)
+	service := &s3DownloadService{
+		client:         client,
+		bucket:         params.Bucket,
+		downloadPath:   params.DownloadPath,
+		numFullRetries: params.NumFullRetries,
+	}
+
+	return service.downloadWithS3Client(ctx, truncatedKeys, logger)
+}
+
+func (service *s3DownloadService) downloadWithS3Client(ctx context.Context, cacheKeys []string, logger log.Logger) (string, error) {
+	var matchedKey string
+	var firstValidKey string
+	err := retry.Times(uint(service.numFullRetries)).Wait(5 * time.Second).TryWithAbort(func(attempt uint) (error, bool) {
+		for _, key := range cacheKeys {
+			fileKey := strings.Join([]string{key, "tzst"}, ".")
+			keyFound, err := service.firstAvailableKey(ctx, fileKey)
+			if err != nil {
+				if errors.Is(errS3KeyNotFound, err) {
+					logger.Debugf("key %s not found in bucket: %s", key, err)
+					continue
+				}
+
+				logger.Debugf("validate key %s: %s", key, err)
+				return err, false
+			}
+
+			matchedKey = keyFound
+			firstValidKey = keyFound
+			return nil, true
+		}
+		return ErrCacheNotFound, true
+	})
+	if err != nil {
+		return "", fmt.Errorf("key validation retries failed: %w", err)
+	}
+
+	err = retry.Times(uint(service.numFullRetries)).Wait(5 * time.Second).TryWithAbort(func(attempt uint) (error, bool) {
+		if err := service.getObject(ctx, firstValidKey); err != nil {
+			return fmt.Errorf("download object: %w", err), false
+		}
+
+		return nil, true
+	})
+	if err != nil {
+		return "", fmt.Errorf("all retries failed: %w", err)
+	}
+
+	return matchedKey, nil
+}
+
+func (service *s3DownloadService) firstAvailableKey(ctx context.Context, key string) (string, error) {
+	_, err := service.client.HeadObject(ctx, &s3.HeadObjectInput{
+		Bucket: aws.String(service.bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		var apiError smithy.APIError
+		if errors.As(err, &apiError) {
+			switch apiError.(type) {
+			case *types.NotFound:
+				return "", errS3KeyNotFound
+			default:
+				return "", fmt.Errorf("aws api error: %w", err)
+			}
+		}
+		return "", fmt.Errorf("generic aws error: %w", err)
+	}
+
+	return key, nil
+}
+
+func (service *s3DownloadService) getObject(ctx context.Context, key string) error {
+	result, err := service.client.GetObject(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(service.bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		return fmt.Errorf("get object: %w", err)
+
+	}
+	defer result.Body.Close() //nolint:errcheck
+	file, err := os.Create(service.downloadPath)
+	if err != nil {
+		return fmt.Errorf("creating file: %w", err)
+	}
+	defer file.Close() //nolint:errcheck
+	body, err := io.ReadAll(result.Body)
```

Comment:
These objects can be quite large (gigabytes+). We should either read the body in fix sized chunks, writing to file simultaneously or use a library that does this for us (the S3 client might have such a method available, I haven't checked).

Without this, the build VM can go out of memory easily.